### PR TITLE
feat(model): Wheel Chocks and GSE Safety Cones

### LINKED ISSUE #3960: Wheel Chocks
**Is your feature request related to a problem? Please describe.**
No. 

**Describe the feature you'd like to see implemented**
It would be great to set Wheel Chocks using the EFB. It would add immersion and it could make the brake cooling process a bit more realistic, because they should cool quicker without the parking brake set. 

**References**
https://en.m.wikipedia.org/wiki/Wheel_chock

**Additional context**
n/a

Discord username (if different from GitHub):
EpicPrimus

### PR DESCRIPTION
**Pull request by bouveng**
Fixes #3960

## Summary of Changes
This PR adds the following

- Adds GSE safety cones to wingtips, engines and tail.
- Adds wheel chocks for nose gear and main landing gears.
- Two new buttons is added to the Ground services page in EFB
- Chocks apply braking force to wheels when active

Deployment is semi-automatic and chocks and safety cones show up given these conditions:

- Chocks and/or Safety Cones are activated in EFB Ground services page.
- `ENG1+2 N1` < 3.5%
- `BEACON LT` is off
- Airplane is on `GND`
- Pushback is disconnected
- Airplane is not moving / rolling

## Screenshots

marketing-cone<p align="right"><i>Ground service equipment safety cone in front of ENG1</i></p>

Microsoft Flight Simulator 2021-12-15 08_43_00<p align="right"><i>Wheel chocks on the left main gear outer wheel.</i></p>

Microsoft Flight Simulator 2021-12-15 08_49_42<p align="right"><i>Manufacturing sticker on chock</i></p>

Screenshot_2021-12-01_at_23 34 56<p align="right"><i>Small manufacturing detail on safety cone</i></p>

Screenshot_2021-12-01_at_23 34 56<p align="right"><i>Small manufacturing detail on wheel chock</i></p>

image <p align="right"><i>EFB settings to enable cones and wheel chocks</i></p>

## References

<p align="right"><i>IATA Ground Operations Manual (IGOM)</i></p>

## Additional context
- ~PR does not contain any logic for actually prevent aircraft from moving, but given conditions above it should be impossible.~ Chocks now prevents plane from being moved.
- ~SDK limitations in terms of positioning of objects relative to datum ref. could make the chocks and cones appear at different heights depending on weights and center of gravity.~ **Fixed by modifying contactpoints/comp. ratio**
- Ropes for the chocks ~will come at a later stage. **_In progress_**~ **Added!**

Discord username (if different from GitHub): @bouveng 

## Testing instructions
1, Spawn C&D
2, Turn on chocks and cones in the EFB on ground services page.
3, Chocks and cones (when enabled) will not be shown given (any of) these conditions:

- `ENG1+2 N1` < 3.5%
- `BEACON LT` is off
- Airplane is on `GND`
- Pushback is connected
- Airplane is moving / rolling

**NOTE!** 
Position of cones and chocks works best on flat surfaces. on tilted / uneven ground they might be misaligned a little and this is nothing we can fix given current SDK.

**TIP!**
To test the chocks braking function. Spawn at the runway at `LFLJ`, then turn off engines, beacon and parking brake.
If chocks are applied, plane should not roll down the hill, if rolling and chocks applied, plane should stop.

## How to download the PR for QA

Every new commit to this PR will cause a new A32NX artifact to be created, built, and uploaded.

1. Make sure you are signed in to GitHub
1. Click on the **Checks** tab on the PR
1. On the left side, click on the bottom **PR** tab
1. Click on the **A32NX** download link at the bottom of the page

### CODE REVIEW

**Comment by crocket63** [comment]
Should I prepare small PR to brake the plane on L:A32NX_MODEL_CONES_HIDDEN info ?

**Comment by crocket63** [comment]
> 

Use case I can think off is that if you empty brake accumulator and you are on a slope, plane would run away through the chokes even with park brake on

**Comment by mehoygg** [comment]
Quality Assurance Tester/Trainee Report

Discord          : mehoygg#1242
Object of testing: #6286
Tier of Testing  : 1
Date             : 02/12/21

Testing Process: 
Spawned C/D in KALX.  
Once I enabled Chocks and Cones, there was x1 Cone at the tip of the wings, a cone under the APU/Tail and in-front of each engine.  Chocks were place under the front and rear of the outermost tire on each side. 
I removed the PRK BRK and the GSE went away.  Set BRKs and GSE returned to their positions. 
I turned on the Beacon and GSE went away and returned to their positions when it was turned off.
To test the N2%, I set APU on, and I set ENG MODE to CRANK and engaged ENG2.  The GSE almost instantly get removed once ENG Master switch is turned ON.  They [GSE] do return to their respective positions once N2% hits 1.0 and lower. 
EFB options work as intended. I removed Cones but kept Chocks and vice versa.  

Negatives:       
N/A
Would windmill affect the GSE if the fans spin too fast if ever?
If ever implemented, thinking of something non related, would it be possible to keep GSE if ever a Air Start option gets introduced?

Testing Results: 
Passed

Conclusions: 
Awesome addition to make it more realistic and great placement of GSE.

Media:       
N/A

**Comment by bouveng** [comment]
> Quality Assurance Tester/Trainee Report

Hi! This is not ready for testing just yet.. but thanks for your report anyhow. =)

/J

**Comment by St54Kevin** [comment]
I've just read through this. Looks great, but I think people will start to ask how to manually set and remove them, especially before pushback. So I would suggest to add a button to the ground equipment page. I would prefer that way over the automatic one hidden in the settings menu.

**Comment by bouveng** [comment]
> So I would suggest to add a button to the ground equipment page.

@St54Kevin 

How would that logic work?
Will you be able to show and hide them in any situation?

- In flight?
- When plane is rolling?
- With engines on?

Explain how you would like it to work.

**Comment by crocket63** [comment]
> > So I would suggest to add a button to the ground equipment page.
> 
> @St54Kevin
> 
> How would that logic work? Will you be able to show and hide them in any situation?
> 
> * In flight?
> * When plane is rolling?
> * With engines on?
> 
> Explain how you would like it to work.

Maybe could be:
"(The current logic to allow it) AND (equipment required on efb)"

**Comment by bouveng** [comment]
@crocket63 

> Maybe could be:
> 
> "(The current logic to allow it) AND (equipment required on efb)"

Elaborate, don't understand what you mean.

You want the current hide to be show, bur still shielded by current logic?

**Comment by crocket63** [comment]
> @crocket63
> 
> > Maybe could be:
> > "(The current logic to allow it) AND (equipment required on efb)"
> 
> Elaborate, don't understand what you mean.
> 
> You want the current hide to be show, but still shielded by current logic?

Yup exactly. If we add a manual action to set chokes, it could be done only if those conditions are OK.  Removing them might be done without conditions.

That's just an idea, lot of other possibilities!

**Comment by St54Kevin** [comment]
I agree to the above. There will be situation in which people will say "I can't remove the chocks before doing my pushback." Because when starting the pushback in most cases there the beacon is still off and engines are not yet running.
Also a big thing is that you have the condition that the park brake is set. But do you know that when on chocks, we often release the parking to reduce wear on the brakes and improve cooling? So when chocks are set, in an optimal world we need a function for the plane to being able to move even with parking brake released.

**Comment by St54Kevin** [comment]
So my suggestion would be to make the following conditions:

- On GND
- ENG's OFF
- BEACON OFF

That way we eliminate the parking brake problem. And I personally would make the auto option defaulting to OFF.

**Comment by bouveng** [comment]
> So my suggestion would be to make the following conditions:
> 
> - On GND
> - ENG's OFF
> - BEACON OFF
> 
> That way we eliminate the parking brake problem. And I personally would make the auto option defaulting to OFF. 

Ok. So people will be able to roll down a hill with cones and chocks following them.

**Comment by cchard360** [comment]
> > So my suggestion would be to make the following conditions:
> > 
> > * On GND
> > * ENG's OFF
> > * BEACON OFF
> > 
> > That way we eliminate the parking brake problem. And I personally would make the auto option defaulting to OFF.
> 
> Ok. So people will be able to roll down a hill with cones and chocks following them.

I believe the suggestion is to have a hidden brake function when the chocks are in place to immobilize the aircraft but allowing people to turn off the parking brake. Just like in the real aircraft. 

However I wonder if the requiment to activate the chocks shouldn't include the parking brake anyway? Isnt that typically SOP? Parking brake on, eng off, beacon off, then the ground crew adds chocks, then you can turn parking brake off?

After they have been placed they should remain in place and the aircraft immobilized until removed by the efb. 

There shouldn't be any requirement to remove them. That would eliminate any issues with people not being able to do so because they did pushback or startup in an unusual order.

**Comment by frankkopp** [comment]
> I agree to the above. There will be situation in which people will say "I can't remove the chocks before doing my pushback." Because when starting the pushback in most cases there the beacon is still off and engines are not yet running.
> Also a big thing is that you have the condition that the park brake is set. But do you know that when on chocks, we often release the parking to reduce wear on the brakes and improve cooling? So when chocks are set, in an optimal world we need a function for the plane not being able to move even with parking brake released. 

For this use case I would suggest to add "Tug connected = false" to the conditions. 
I'm not concerned about a manual option if the other options are logical. In rl they would be removed be ground crew based on these other "hints" (beacon, pushback, etc.)

**Comment by crocket63** [comment]
> I believe the suggestion is to have a hidden brake function when the chocks are in place to immobilize the aircraft but allowing people to turn off the parking brake. Just like in the real aircraft.
> 

This is available here: https://github.com/flybywiresim/a32nx/pull/6307

**Comment by Saschl** [comment]
Button fixed available here: https://github.com/Saschl/a32nx/tree/chocks-gs

The state only depends on the setting now, not on the actual visibility of the models.

File: src/systems/a320_systems/src/hydraulic.rs [hunk 1]
```diff
@@ -1572,6 +1580,8 @@ impl A320BrakingForce {
         self.right_braking_force = self.right_braking_force.max(0.).min(1.);
 
         self.correct_with_flaps_state(context);
+
+        self.apply_chocks_braking();
```
Comment (crocket63) [inline src/systems/a320_systems/src/hydraulic.rs:1584]: I should have probably better named this "update_chocks_braking"

Comment (bouveng) [reply src/systems/a320_systems/src/hydraulic.rs:1584]: @crocket63 I can change that.

Comment (bouveng) [reply src/systems/a320_systems/src/hydraulic.rs]: @crocket63 Changed to `update_chocks_braking`

**Comment by frankkopp** [comment]
@bouveng any idea when this will be merged?

**Comment by bouveng** [comment]
> @bouveng any idea when this will be merged?

Hi man! It is 99% done. I just want to try out a final thing, with regards to making the chock nodes children of something else than being referenced from datum reference. Just have not had time to try this yet.

Br.

**Comment by matanbr** [comment]
A

**Comment by TJCAviation** [comment]
looking forward to this! Any news on a merge or how to test this PR?

**Review by svengcz — APPROVED** [review]
LGTM

**Comment by mico975** [comment]
Quality Assurance Tester Report

Discord          : mico#3145
Object of testing: #6286
Tier of Testing  : 1
Date             : 11/05/2022

Testing Process: 
Deployment is semi-automatic and chocks and safety cones show up given these conditions:

Chocks and/or Safety Cones are activated in EFB Ground services page. - **OK**
ENG1+2 N1 < 3.5% - **OK**
BEACON LT is off - **OK**
Airplane is on GND - **OK**
Pushback is disconnected - **OK**
Airplane is not moving / rolling - **OK**

Negatives:       
N/A

Testing Results: 
Passed

Conclusions: 
LGTM

Media:       
image
image

**Comment by Letreivchen** [comment]
Quality Assurance Tester Report

Discord          : Letreiv#9404
Object of testing: #6286 
Tier of Testing  : 1
Date             : 11.05.2022

Testing Process: 
1, Spawned C&D
2, Turned on chocks and cones in the EFB on ground services page. ✔
3, Chocks and cones (when enabled) will not be shown given (any of) these conditions:

- ENG1+2 N1 < 3.5% ✔
- BEACON LT is off ✔
- Airplane is on GND ✔
- Pushback is connected ✔
- Spawned in LFLJ to check:
that the Airplane is moving / rolling without the chocks and standing still with chocks enabled ✔

Negatives:       
As stated in the PR description there are some viusal issues on uneven/tilted surfaces like the runway in LFLJ

Testing Results: 
PASSED

Conclusions: 
Great addition and long awaited feature, glad to see it finished 💯 

Media:

**Review by Benjozork — APPROVED** [review]

**Comment by ewold** [comment]
Quality Assurance Trainee Report

Discord: E-wold#8195
Object of testing: https://github.com/flybywiresim/a32nx/pull/6286
Tier of Testing : 1
Date : 11.05.2022

Testing Process:
1, Spawned C&D
2, Turned on chocks and cones in the EFB on ground services page. ✔
3, Chocks and cones (when enabled) will not be shown given (any of) these conditions:

ENG1+2 N1 < 3.5% ✔
BEACON LT is off ✔
Airplane is on GND ✔
Pushback is connected ✔
Spawned in LFLJ to check:
that the Airplane is moving / rolling without the chocks and standing still with chocks enabled ✔

Negatives:
 
Testing Results: PASSED

Conclusions: Looks good , EFB works , didnt test it in LFLJ due to time constraint
fbwg1

fbwg2
fbwg3
fbwg4

**Pull request merged**